tidy3d/components/scene.py

```diff
@@ -0,0 +1,1171 @@
+    >>> sim = Scene(
+    ...     size=(3.0, 3.0, 3.0),
+    ...     structures=[
+    ...         Structure(
+    ...             geometry=Box(size=(1, 1, 1), center=(0, 0, 0)),
+    ...             medium=Medium(permittivity=2.0),
+    ...         ),
+    ...     ],
+    ...     medium=Medium(permittivity=3.0),
+    ... )
+    """
+
+    medium: MediumType3D = pydantic.Field(
+        Medium(),
+        title="Background Medium",
+        description="Background medium of simulation, defaults to vacuum if not specified.",
+        discriminator=TYPE_TAG_STR,
+    )
+
+    structures: Tuple[Structure, ...] = pydantic.Field(
+        (),
+        title="Structures",
+        description="Tuple of structures present in simulation. "
+        "Note: Structures defined later in this list override the "
+        "simulation material properties in regions of spatial overlap.",
+    )
+
+    """ Validating setup """
+
+    _structures_in_bounds = assert_objects_in_sim_bounds("structures", error=False)
+
+    # make sure all names are unique
+    _unique_structure_names = assert_unique_names("structures")
+
+    @pydantic.validator("structures", always=True)
+    def _validate_num_mediums(cls, val):
+        """Error if too many mediums present."""
+
+        if val is None:
+            return val
+
+        mediums = {structure.medium for structure in val}
+        if len(mediums) > MAX_NUM_MEDIUMS:
+            raise SetupError(
+                f"Tidy3d only supports {MAX_NUM_MEDIUMS} distinct mediums."
+                f"{len(mediums)} were supplied."
+            )
+
+        return val
+
+    @pydantic.validator("structures", always=True)
+    def _structures_not_at_edges(cls, val, values):
+        """Warn if any structures lie at the scene boundaries."""
+
+        if val is None:
+            return val
+
+        sim_box = Box(size=values.get("size"), center=values.get("center"))
+        sim_bound_min, sim_bound_max = sim_box.bounds
+        sim_bounds = list(sim_bound_min) + list(sim_bound_max)
+
+        with log as consolidated_logger:
+            for istruct, structure in enumerate(val):
+                struct_bound_min, struct_bound_max = structure.geometry.bounds
+                struct_bounds = list(struct_bound_min) + list(struct_bound_max)
+
+                for sim_val, struct_val in zip(sim_bounds, struct_bounds):
+
+                    if isclose(sim_val, struct_val):
+                        consolidated_logger.warning(
+                            f"Structure at structures[{istruct}] has bounds that extend exactly to "
+                            "simulation edges. This can cause unexpected behavior. "
+                            "If intending to extend the structure to infinity along one dimension, "
+                            "use td.inf as a size variable instead to make this explicit."
+                        )
+
+        return val
+
+    """ Post-init validators """
+
+    def _post_init_validators(self) -> None:
```

Comment:
do we need this here? or should it just be in `Simulation`? I forget how this works or if we need it to be defined inside of a scene itself.